### PR TITLE
New Edit Cycle form load architecture and UI for "Details" section (Section 1/4)

### PR DESCRIPTION
**Pull request by johnnyd-eth**
**Code overview:**
- Loads an Ant-D form into a react context.
- Loads project FC data from redux, converts to formData object and passes into `<Form` with `initialValues`.
- Where necessary, updates form values from the context.
- New UI components and UI complete for the `Details` section (Section 1/4) (fixes JB-546)

**UI:**

https://github.com/jbx-protocol/juice-interface/assets/96150256/8be1ffaf-a8bd-4707-b518-45539d20ff1a

### CODE REVIEW

File: src/components/inputs/JuiceListbox.tsx [hunk 1]
```diff
@@ -9,6 +9,7 @@ interface JuiceListboxOption<T> {
 }
 
 interface JuiceListboxProps<T> {
+  disabled?: boolean
```
Comment (johnnyd-eth) [inline src/components/inputs/JuiceListbox.tsx:12]: Make `JuiceListboxProps` support `disabled` prop. Shows `N.A` as value when disabled

File: src/components/inputs/JuiceSwitch.tsx [hunk 2]
```diff
@@ -5,12 +5,13 @@ import { classNames } from 'utils/classNames'
 
 export const JuiceSwitch = ({
   label,
+  description,
   value,
   onChange,
-}: FormItemInput<boolean> & { label?: ReactNode }) => {
-  const switchId = `switch-${label}`
+}: FormItemInput<boolean> & { label?: ReactNode; description?: ReactNode }) => {
+  const switchId = `switch-${label}-${Math.random()}`
```
Comment (johnnyd-eth) [inline src/components/inputs/JuiceSwitch.tsx:12]: Bug caused when the same component was used as a `label` (e.g. `<TooltipLabel />`)

File: src/components/inputs/JuiceSwitch.tsx [hunk 3]
```diff
@@ -20,9 +21,14 @@ export const JuiceSwitch = ({
...
       />
       {label && (
-        <label htmlFor={switchId} className="text-black dark:text-grey-200">
-          {label}
-        </label>
+        <div>
+          <label htmlFor={switchId} className="text-black dark:text-grey-200">
+            {label}
+          </label>
+          {description ? (
```
Comment (johnnyd-eth) [inline src/components/inputs/JuiceSwitch.tsx:28]: Support a description in JuiceSwitch, note both description and non-description instances look fine:

Screen Shot 2023-07-16 at 12 39 43 pm

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/ProjectSettingsLayout.tsx [hunk 4]
```diff
@@ -14,7 +14,7 @@ export const ProjectSettingsLayout: React.FC<React.PropsWithChildren> = ({
 
   return (
     <>
-      <header className="sticky top-0 right-0 mb-8 border-b border-solid border-b-grey-100 bg-white dark:border-b-slate-500 dark:bg-slate-900">
+      <header className="sticky top-0 right-0 z-50 mb-8 border-b border-solid border-b-grey-100 bg-white dark:border-b-slate-500 dark:bg-slate-900">
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/ProjectSettingsLayout.tsx:17]: Fixes JB-264 (https://linear.app/peel/issue/JB-624/overlapping-button)

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/V2V3SettingsProvider.tsx [hunk 5]
```diff
@@ -19,7 +20,9 @@ export const V2V3SettingsProvider: React.FC<React.PropsWithChildren> = ({
     <AppWrapper hideNav>
       <Provider store={store}>
         <V2V3ProjectPageProvider projectId={projectId}>
-          <TransactionProvider>{children}</TransactionProvider>
+          <TransactionProvider>
+            <EditCycleFormProvider>{children}</EditCycleFormProvider>
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/V2V3SettingsProvider.tsx:24]: Use this context so we can update the form instance without passing through as props everywhere

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCycleFormContext.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+export const useEditCycleForm = () => useContext(EditCycleDataContext)
+
+export const EditCycleFormProvider: React.FC<
+  React.PropsWithChildren<unknown>
+> = ({ children }) => {
+  const { initialFormData, editCycleForm } = useLoadEditCycleData()
+
+  return (
+    <EditCycleDataContext.Provider value={{ initialFormData, editCycleForm }}>
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCycleFormContext.tsx:24]: Context also keeps track of the initial data

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCycleFormFields.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+export type DetailsSectionFields = {
+  duration: number
+  durationUnit: DurationOption
+  ballot: string
+  allowSetTerminals: boolean
+  allowSetController: boolean
+  pausePay: boolean
+}
+
+export type PayoutsSectionFields = {
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCycleFormFields.ts:15]: Prep for next PRs

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCyclePage.tsx [hunk 8]
```diff
@@ -1,37 +1,22 @@
...
+import { Button, Form } from 'antd'
 import ExternalLink from 'components/ExternalLink'
+import Loading from 'components/Loading'
 import { ProjectMetadataContext } from 'contexts/shared/ProjectMetadataContext'
 import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
 import Link from 'next/link'
 import { useContext } from 'react'
 import { helpPagePath, settingsPagePath } from 'utils/routes'
-
-function EditCycleSection({
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/EditCyclePage.tsx:10]: Moved to its own component

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/hooks/LoadEditCycleData.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+export const useLoadEditCycleData = () => {
+  const [initialFormData, setInitialFormData] = useState<
+    EditCycleFormFields | undefined
+  >(undefined)
+  const { initialEditingData } = useInitialEditingData({ visible: true })
+  const [editCycleForm] = Form.useForm<EditCycleFormFields>()
+
+  useEffect(() => {
+    // Extracts EditCycleFormFields from initialEditingData (which comes from redux)
+    if (initialEditingData) {
```
Comment (johnnyd-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/hooks/LoadEditCycleData.tsx:31]: :pain:

File: src/constants/v2v3/ballotStrategies/index.ts [hunk 10]
```diff
@@ -102,14 +102,14 @@ export function ballotStrategiesFn(network?: NetworkName): BallotStrategy[] {
     // v2, hacky, sorry
     {
       id: 'threeDayV2',
-      name: t`3-day deadline`,
+      name: t`V2 3-day deadline`,
```
Comment (johnnyd-eth) [inline src/constants/v2v3/ballotStrategies/index.ts:105]: Bug on prod, currently two sets of duplicate labels on deadline dropdown

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/DetailsSection/DurationFields.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import { useEffect, useState } from 'react'
+import { useEditCycleForm } from '../EditCycleFormContext'
+
+export function DurationFields() {
+  const [durationEnabled, setDurationEnabled] = useState<boolean>()
+
+  const { editCycleForm, initialFormData } = useEditCycleForm()
+
+  // Set initial value of Switch
+  useEffect(() => {
```
Comment (wraeth-eth) [inline src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/DetailsSection/DurationFields.tsx:16]: avoid useEffect for state changes if you can - very error prone and creates unintentional bugs

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/hooks/LoadEditCycleData.tsx (hunk 9 above)
Comment (wraeth-eth) [reply src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/hooks/LoadEditCycleData.tsx:31]: Yep it looks like it 😅

**Comment by wraeth-eth** [comment]
When I open, nfts is already highlighted despite doing nothing
image

http://localhost:3000/v2/p/1/settings/cycle

**Comment by wraeth-eth** [comment]
image
Is the formatting difference between duration and deadline expected - i.e, the colors?

**Comment by wraeth-eth** [comment]
image
borders on dark mode are wrong color

**Comment by johnnyd-eth** [comment]
> image Is the formatting difference between duration and deadline expected - i.e, the colors?

This is because of the input min value of 1. I agree it's a bit jarring so I've changed min-value to 0 so the red doesn't show

File: src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/DetailsSection/DurationFields.tsx (hunk 11 above)
Comment (johnnyd-eth) [reply src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/NewEditCyclePage/DetailsSection/DurationFields.tsx]: Yep this was unnecessary, worked the initial value into the state definition now and it's working fine

**Comment by johnnyd-eth** [comment]
> image borders on dark mode are wrong color

Fixed:
Screen Shot 2023-07-17 at 10 30 29 am

**Review by wraeth-eth — APPROVED** [review]

**Pull request merged**